FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,425 @@
+              it.copy(
+                forwardedFolder = "Outbox",
+                forwardedUid = Random.nextLong(),
+                decryptWhenForward = true
+              )
+            }
+
+            //create MimeMessage from content + attachments
+            val mimeMsgWithAttachments = EmailUtil.createMimeMsg(
+              context = applicationContext,
+              sess = Session.getDefaultInstance(Properties()),
+              account = account,
+              msgEntity = msgEntity.copy(isEncrypted = false),
+              atts = attachments
+            )
+
+            //get recipients that will be used to create password-encrypted msg
+            val toCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.TO,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.TO)
+              )
+            )
+
+            val ccCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.CC,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.CC)
+              )
+            )
+
+            val bccCandidates = mimeMsgWithAttachments.getMatchingRecipients(
+              type = Message.RecipientType.BCC,
+              list = GeneralUtil.getRecipientsWithoutUsablePubKeys(
+                context = applicationContext,
+                emails = mimeMsgWithAttachments.getAddresses(Message.RecipientType.BCC)
+              )
+            )
+
+            if (toCandidates.isNotEmpty()) {
+              //start of creating and uploading a password-protected msg to FES
+              val fromAddress = mimeMsgWithAttachments.getFromAddress()
+              val domain = EmailUtil.getDomain(fromAddress)
+              val idToken = getGoogleIdToken()
+              val replyToken = fetchReplyToken(apiRepository, domain, idToken)
+              val messageUploadRequest = MessageUploadRequest(
+                associateReplyToken = replyToken,
+                from = fromAddress,
+                to = toCandidates.map { (it as InternetAddress).address },
+                cc = ccCandidates.map { (it as InternetAddress).address },
+                bcc = bccCandidates.map { (it as InternetAddress).address }
+              )
+
+              //prepare bodyWithReplyToken
+              val replyInfo = Base64.getEncoder().encodeToString(
+                GsonBuilder().create().toJson(messageUploadRequest).toByteArray()
+              )
+
+              val infoDiv = genInfoDiv(replyInfo)
+              val originalText = getDecryptedContentFromMessage(
+                mimeMsgWithAttachments,
+                accountSecretKeys,
+                keysStorage
+              )
```

Comment:
Why not get `originalText` from `msgEntity`? It seems it's being re-parsed unnecessarily.

Reply:
We can have 2 ways:

### 1. we can do all thing simply. 

**pros**

- things will look much simpler and I will work with original source

**cons**

- we will wait while the current message will be sent

### 2. we can do all thing in the background tasks. 

**pros**

- a user will not wait on the compose screen and will be able to iteract with the app

**cons**

- things looks a little difficult

Currently we have the second way. Let's me explain how it works

1. A user create a password protected message. We add `OutgoingMessageInfo` to a queue to `PrepareOutgoingMessagesJobIntentService`
2. `PrepareOutgoingMessagesJobIntentService` creates en encrypted message with 'message.asc' pattern. In the current PR we add only a few changes to this class. It sounds good, we should not break the existing logic. `PrepareOutgoingMessagesJobIntentService` adds to the local database new outgoing message with encrypted `password`(using AndroidKeyStore) and enqueue a new task for `HandlePasswordProtectedMsgWorker` if there are no forwarded attachments or `ForwardedAttachmentsDownloaderWorker` if we have them
3. `HandlePasswordProtectedMsgWorker` fetches password-protected messages from the local database and tries to upload them to FES. At this place we receive an encrypted message (MIME message with `message.asc` + attachments). So to construct a plain MIME message we have to decrypt `message.asc`.

Reply:
I see. Ok